Refactor and improvements to HttpContent Factories

lachlan-roberts opened the pull request:
- flatten cache in HttpCachingHttpContentFactory
- remove old CachedContentFactory
- HttpContent uses RetainableByteBuffer
- use buffer pool in CachingHttpContentFactory
- add configuration to reduce/prevent file system accesses for cached HttpContent
- add separate HttpContent factory for file mapping

lachlan-roberts committed:
Refactor for CachingContentFactory and ResourceService

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
only check file modified time once per second for cached files

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix stampede creating CachingHttpContent & remove usage of getPreCompressedContentFormats

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
remove usage of HttpContent.getPreCompressedContentFormats for jetty-core & ee10

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
remove getPreCompressedContentFormats from HttpContent

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
remove getPrecompressedContents from HttpContent and ee9 impl

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
use jetty-core ResourceContentFactory in ee9

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
replace use of CachedContentFactory with CachingContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups of HttpContentWrapper and ResourceService

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
extract the generation of getCompressedContentFormats to separate contentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cache more fields in CachingHttpContent

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
work in progress to improve CachingContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
further work on CachingContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

lachlan-roberts committed:
fix failing test cases for ResourceService and Etags

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Rename all HttpContent factories

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups of HttpContent factories

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
improve configuration options for HttpContent.Factory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
separate evicting logic from CachingHttpContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
use RetainableByteBuffer for getBuffer in HttpContent

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix bugs in CachingHttpContentFactory, ResourceService and FileMappedHttpContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
ResourceHandlerTest should validate cache entry on every request

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cannot add ContentFactory to servletContext attribute as there may be multiple factories in a single context

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

lachlan-roberts committed:
use name annotation for EvictingCachingContentFactory constructors

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented:
@lachlan-roberts are those test failures flakes?

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+ * {@link HttpContent.Factory} implementation that wraps any other {@link HttpContent.Factory} instance
+ * using it as a caching authority.
+ * Only HttpContent instances whose path is not a directory are cached.
+ */
```
You need to document shrink and evict policy (even if delegated to another class)

lachlan-roberts replied:
Done.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+        long len = httpContent.getContentLengthValue();
+        if (len <= 0)
+            return false;
+        if (httpContent instanceof FileMappedHttpContentFactory.FileMappedContent)
+             return true;
```
This is very specific.   Perhaps a tag interface would be better?

lachlan-roberts replied:
Not sure what you mean by this.

lachlan-roberts replied:
Add method on `HttpContent` called `boolean getBufferOccupancy()`.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+    private void shrinkCache()
+    {
+        // While we need to shrink
+        while (_cache.size() > 0 && (_cache.size() > _maxCachedFiles || _cachedSize.get() > _maxCacheSize))
```
get `_cache.size()` only once and assign to variable.

lachlan-roberts replied:
Ok sure, but will also need to reassign for each iteration of the loop.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+                return cachingContent;
+            });
+
+            if (wasAdded.get())
```
Is it worthwhile having the `wasAdded` atomic?  Why not just call `shrinkCache` anyway and that will do a volitile lookup of the size, which is the same cost.

lachlan-roberts replied:
`shrinkCache` will have to do 2 volatile lookups, `_cache.size()` and `_cachedSize.get()`.

Ideally we wouldn't need the AtomicBoolean here, it is just because we can't know if `computeIfAbsent` created a new entry or not.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+            }
+            _etagField = etagField;
+
+            // Map the content into memory if possible.
```
```suggestion
            // Read the content into memory if possible.
```

What if the buffer was a `FileMappedHttpContentFactory.FileMappedContent`, wont you also read it?

lachlan-roberts replied:
Yes, but I don't think we should be reading it. The check in `CachingContentFactory.isCacheable`
```java
if (httpContent instanceof FileMappedHttpContentFactory.FileMappedContent)
             return true;
```
indicates that we don't count these towards the size so we shouldn't be copying it into indirect buffer.

I will change this to 
```java
// Read the content into memory if the HttpContent does not already have a buffer.
RetainableByteBuffer buffer = httpContent.getBuffer();
if (buffer != null)
{
    ...
```

joakime commented:
The `TryPathsHandlerTest` failures seem to indicate a change in either Request.Processor behavior, or ResourceHandler behavior.
Not a flaky test.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,540 @@
+            _lastAccessed.set(nanosTime);
+        }
+
+        @Override
+        public HttpField getContentType()
+        {
+            return null;
+        }
+
+        @Override
+        public String getContentTypeValue()
+        {
+            return null;
+        }
+
+        @Override
+        public String getCharacterEncoding()
+        {
+            return null;
+        }
+
```
I think these should all be default impls on the interface.

gregw replied:
and most of the other methods in this class

lachlan-roberts replied:
Why? this is the only implementation which ever returns null and it is only an internal implementation which is never returned by `getContent()`.

gregw replied:
Because:
 + in future there may be other implementations.
 + this is an impl that actually does something, so best to not crowd the code with boiler plate.  Having the null impl in the interface is a better place that crowding this file.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EvictingCachingContentFactory.java:
```diff
@@ -0,0 +1,165 @@
+ * elapsed the entry will be invalid and will be evicted from the cache at the next access.
+ * </p>
+ */
+public class EvictingCachingContentFactory extends CachingHttpContentFactory implements Runnable
```
The naming is not exactly correct because it is not this class that does the evicting. That is done by `CachingHttpContentFactory` if the content is not valid.    Perhaps `ValidatingCachingContentFactory`?

lachlan-roberts replied:
Done.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EvictingCachingContentFactory.java:
```diff
@@ -0,0 +1,165 @@
+
+    public EvictingCachingContentFactory(@Name("authority") HttpContent.Factory authority,
+                                         @Name("validationTime") long validationTime,
+                                         @Name("scheduler") Scheduler scheduler,
```
need to better document the optional sweeper

lachlan-roberts replied:
Done.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EvictingCachingContentFactory.java:
```diff
@@ -0,0 +1,165 @@
+            else if (!value.isValid())
+                removeFromCache(value);
+        }
+        schedule();
```
maybe put in a try-finally... just in case.

lachlan-roberts replied:
Done.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,72 @@
+                    LOG.debug("Error getting Mapped Buffer", t);
+            }
+
+            return super.getBuffer();
```
This feels wrong.  If the resource cannot be mapped, then we should not create a `FileMappedContent`.

joakime replied:
Here's some cases where BufferUtil.toMappedBuffer(Path) will not work.

* Extremely large files (I'm talking huge, like 200GB+)
* Real OS FileSystems based on dynamic content. (filesystems that have content that doesn't exist until it's read from, and can be different with each read, many such things exist on Windows)
* Emulated OS FileSystems. (eg: a Windows machine, in the WSL layer, reading a windows drive via the WSL layer)
* Some OS based Network FileSystems.
* Java FileSystems based on virtual concepts (eg:zipfs)

lachlan-roberts replied:
Is there a test we can do to see if it is mappable? If so this should be done in `getContent()`.

Or I could try and call `BufferUtil.toMappedBuffer` in `getContent()` and only wrap if this does not throw. But if the file changes on disk it could become unmappable even after this `FileMappedContent` has been created.

If we cache a mapped buffer the size of the buffer could change if the file is changed on disk, so should we not cache any `FileMappedContent`.

joakime replied:
I think the easiest way, in a testcase, is to have a zipfs (mounted jar) and try to use BufferUtil.toMappedBuffer(Path), which should fail in 100% of use cases.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,72 @@
+        {
+            try
+            {
+                ByteBuffer buffer = BufferUtil.toMappedBuffer(content.getResource().getPath());
```
Isn't this re-mapping the buffer on every call to getBuffer()?  Surely it should be mapped already and just sliced?

joakime replied:
Here's some cases where BufferUtil.toMappedBuffer(Path) will not work.

* Extremely large files (I'm talking huge, like 200GB+)
* Real OS FileSystems based on dynamic content. (filesystems that have content that doesn't exist until it's read from, and can be different with each read, many such things exist on Windows)
* Emulated OS FileSystems. (eg: a Windows machine, in the WSL layer, reading a windows drive via the WSL layer)
* Some OS based Network FileSystems.
* Java FileSystems based on virtual concepts (eg:zipfs)

gregw reviewed (dismissed):
Only a partial review so far.

lachlan-roberts committed:
changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

lachlan-roberts commented:
@gregw @joakime I assumed `TryPathsHandlerTest` was flakey because it was not failing in this current branch, but looks like there are some new changes in `12.0.x` which cause it to fail with this PR.

I will investigate.

lachlan-roberts committed:
do not cache notFound entries in TryPathsHandlerTests ResourceHandler

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
It was failing because the `ResourceHandler` in the `TryPathsHandlerTest` was caching the "not found" entries. So when they were created they were still reported as not found.

lachlan-roberts commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,72 @@
+    public HttpContent getContent(String path) throws IOException
+    {
+        HttpContent content = _factory.getContent(path);
+        if (content != null && content.getContentLengthValue() > 16 * 1024)
```
This `16 * 1024` value should be configurable.

lachlan-roberts committed:
fix release in CachingHttpContentFactory and always use ByteBufferPool

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
add getBytesOccupied method to HttpContent

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix javadoc in ValidatingCachingContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,110 @@
+
+    public FileMappedHttpContentFactory(HttpContent.Factory factory)
+    {
+        this(factory, 16 * 1024);
```
This constant should be javadoc'ed.

lachlan-roberts replied:
done

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,110 @@
+        return content;
+    }
+
+    public static class FileMappedContent extends HttpContentWrapper
```
This should be `private` or maybe package-protected.

lachlan-roberts replied:
I made this private now

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -38,7 +38,7 @@ public class RetainableByteBuffer extends Retainable.ReferenceCounter
     private final AtomicLong lastUpdate = new AtomicLong(NanoTime.now());
 
-    RetainableByteBuffer(ByteBuffer buffer, Consumer<RetainableByteBuffer> releaser)
+    public RetainableByteBuffer(ByteBuffer buffer, Consumer<RetainableByteBuffer> releaser)
```
It doesn't look like that is still needed.

lachlan-roberts replied:
reverted

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/TryPathsHandlerTest.java:
```diff
@@ -94,7 +96,16 @@ public void dispose()
+            @Override
+            protected HttpContent.Factory setupContentFactory()
+            {
+                // We don't want to cache not found entries for this test.
```
There should be a more straightforward way to achieve this. Perhaps a config setting?

lachlan-roberts replied:
I think there would be too many settings to introduce config for.
And this doesn't have init parameters like the `DefaultServlet`.

sbordet replied:
I would have expected `ResourceHandler` to have a `setHttpContentFactory(HttpContent.Factory)` so that here I could have set just the `ResourceHCF`.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -595,6 +595,21 @@ public static void readFrom(Path path, ByteBuffer buffer) throws IOException
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(byteBuffer, pos);
```
Why flipping in a `finally` block?

lachlan-roberts replied:
I think Greg suggested this in previous review because `channel.read` can throw.

gregw replied:
Thinking about it, it doesn't matter either way. If the method throws, the buffer cannot be trusted.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -595,6 +595,21 @@ public static void readFrom(Path path, ByteBuffer buffer) throws IOException
+    public static void readFrom(ReadableByteChannel newReadableByteChannel, long contentLengthValue, ByteBuffer byteBuffer) throws IOException
+    {
+        int pos = BufferUtil.flipToFill(byteBuffer);
+        try (ReadableByteChannel channel = newReadableByteChannel)
```
This is going to close the `ReadableByteChannel`, and this should not happen here: the usual convention is to always close an IO resource in the same code context it was opened.

lachlan-roberts replied:
Good point, I will move this try with resources to where this method is used.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -595,6 +595,21 @@ public static void readFrom(Path path, ByteBuffer buffer) throws IOException
+        try (ReadableByteChannel channel = newReadableByteChannel)
+        {
+            int read = 0;
+            while (read != contentLengthValue)
```
This loop might spin. Not only should it call `Thread.onSpinWait()` when it does (i.e.: when `read` returns 0) but that should also be clearly stated in the method's javadoc.

lachlan-roberts replied:
I have updated this loop/javadoc also putting in a case for (read < 0) so please re-review.

lachlan-roberts replied:
I have updated this loop/javadoc also putting in a case for (read < 0) so please re-review.

lorban commented on jetty-ee9/jetty-ee9-servlet/src/main/java/org/eclipse/jetty/ee9/servlet/DefaultServlet.java:
```diff
@@ -308,6 +300,17 @@ public void init()
             LOG.debug("resource base = {}", _baseResource);
     }
 
+    private static ByteBufferPool getByteBufferPool(ContextHandler contextHandler)
```
I think I saw the same piece of code duplicated in other classes. Can't it be factored out?

lachlan-roberts replied:
Don't thinks so, its duplicated across ee9 and ee10 which also use different `ContextHandler` classes.

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,571 @@
+    private final ConcurrentMap<String, CachingHttpContent> _cache = new ConcurrentHashMap<>();
+    private final AtomicLong _cachedSize = new AtomicLong();
+    private final ByteBufferPool _byteBufferPool;
+    private int _maxCachedFileSize = 128 * 1024 * 1024;
```
Those default values should be javadoc'ed.

lachlan-roberts replied:
done

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,571 @@
+        private final HttpField _contentLength;
+        private final Instant _lastModifiedInstant;
+        private final HttpField _lastModified;
+        private final AtomicInteger _referenceCount = new AtomicInteger(1);
```
You probably should use `Retainable.ReferenceCounter` instead.

lachlan-roberts replied:
done

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,571 @@
+        private final String _cacheKey;
+        private final HttpField _etagField;
+        private final long _contentLengthValue;
+        private final AtomicLong _lastAccessed = new AtomicLong();
```
Looks like a `volatile long` is all you need.

lachlan-roberts replied:
done

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,571 @@
+
+    protected static class NotFoundHttpContent implements CachingHttpContent
+    {
+        private final AtomicLong _lastAccessed = new AtomicLong();
```
Looks like a `volatile long` is all you need.

lachlan-roberts replied:
done

lachlan-roberts committed:
changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,124 @@
+    public HttpContent getContent(String path) throws IOException
+    {
+        HttpContent content = _factory.getContent(path);
+        if (content != null && content.getContentLengthValue() > _minFileSize)
```
Should have a check about max size too (cannot be greater than Integer.MAX_VALUE as per `FileChannel.map()` javadocs).

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,124 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileMappedHttpContentFactory implements HttpContent.Factory
```
I'd call this class `FileMappingHttpContentFactory`.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpContentWrapper.java:
```diff
@@ -24,13 +25,18 @@
  */
 public class HttpContentWrapper implements HttpContent
```
Can we move this class as an inner class of `HttpContent`? To be consistent with all other wrappers we have.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,124 @@
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            try (AutoLock lock = _lock.lock())
```
At least let's use double-checked locking.

lachlan-roberts replied:
Changed it to do that now, I had to introduce a `SENTINEL_BUFFER` static final field. And remove the `_processed` boolean.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,124 @@
+        @Override
+        public long getBytesOccupied()
+        {
+            try (AutoLock lock = _lock.lock())
```
Don't duplicate code, call `getByteBuffer()` instead.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpContentWrapper.java:
```diff
@@ -110,20 +122,26 @@ public Resource getResource()
 
     @Override
-    public Map<CompressedContentFormat, ? extends HttpContent> getPrecompressedContents()
+    public ByteBuffer getByteBuffer()
```
Missing override for `getBytesOccupied()`.

lachlan-roberts replied:
fixed

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java:
```diff
@@ -0,0 +1,586 @@
+                    long contentLengthValue = getContentLengthValue();
+                    if (contentLengthValue > _maxCachedFileSize)
+                    {
+                        byteBuffer = _byteBufferPool.acquire((int)contentLengthValue, false);
```
This cast to `int` should be handled better, as we don't want to blindly cast, but check if `contentLengthValue < Integer.MAX_VALUE` and only then cache it.

lachlan-roberts replied:
I think this check should instead be `(contentLengthValue <= _maxCachedFileSize)`, and `_maxCachedFileSize` is an `int` so it will be <= `Integer.MAX_VALUE`.

But we also do this check in `isCacheable` so there should be no way to get here with a contentLengthValue greater than `Integer.MAX_VALUE`.

sbordet replied:
I would not restrict `maxCachedFileSize` to an `int`, given there may be systems with a lot of memory.
A 2 GiB file is not that large these days.
This is even more so because `getContentLengthValue()` is a `long` already.

lachlan-roberts replied:
So you're saying if the `maxCachedFileSize > Integer.MAX_VALUE` the entry should be added to the cache but there should be no buffer in the cache? So we will be caching `lastModified` `contentLength` etc.. but not caching the content.

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ResourceHttpContentFactory.java:
```diff
@@ -82,29 +72,7 @@ private HttpContent load(String pathInContext, Resource resource)
         if (resource.isDirectory())
             return new ResourceHttpContent(resource, _mimeTypes.getMimeByExtension(resource.toString()));
```
`Resource.toString()` is not implemented in `Resource`.
I would not assume "oh, but this is a PathResource so toString() is implemented`.
Plus `toString()` may be changed to a different format and totally break this code.
Why can it not be `pathInContext` like the line below?

lachlan-roberts replied:
Not sure, this is pre-existing logic which has been maintained here.

So I'm not sure why directories need to be handled differently here, there is `resource.getFileName()` which is better than the `toString()`, but I also think `pathInContext` is good enough.

lachlan-roberts replied:
I have now removed this the `resource.isDirectory` block.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ValidatingCachingContentFactory.java:
```diff
@@ -0,0 +1,193 @@
+                                           @Name("validationTime") long validationTime,
+                                           @Name("byteBufferPool") ByteBufferPool byteBufferPool,
+                                           @Name("scheduler") Scheduler scheduler,
+                                           @Name("sweepDelay") long sweepDelay,
```
Rename to `sweepPeriod`.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ValidatingCachingContentFactory.java:
```diff
@@ -0,0 +1,193 @@
+                                           @Name("byteBufferPool") ByteBufferPool byteBufferPool,
+                                           @Name("scheduler") Scheduler scheduler,
+                                           @Name("sweepDelay") long sweepDelay,
+                                           @Name("maxCacheIdleTime") long maxCacheIdleTime)
```
Rename to `idleTimeout` like everywhere else.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -65,15 +67,15 @@
+    private final List<CompressedContentFormat> _precompressedFormats = new ArrayList<>();
+    private final Map<String, List<String>> _preferredEncodingOrderCache = new ConcurrentHashMap<>();
+    private final List<String> _preferredEncodingOrder = new ArrayList<>();
 
```
Remove newline.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -101,34 +103,55 @@ public Resource getStylesheet()
     }
 
-    public HttpContent.ContentFactory getContentFactory()
+    public HttpContent.Factory getContentFactory()
```
Rename to `getHttpContentFactory()`.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -101,34 +103,55 @@ public Resource getStylesheet()
     }
 
-    public void setContentFactory(HttpContent.ContentFactory contentFactory)
+    public void setContentFactory(HttpContent.Factory contentFactory)
```
Rename to `setHttpContentFactory()`.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -61,31 +67,49 @@ public ResourceHandler()
+        return (byteBufferPool == null) ? new NoopByteBufferPool() : byteBufferPool;
+    }
+
+    protected HttpContent.Factory setupContentFactory()
```
Rename to `setupHttpContentFactory()`.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -120,7 +144,7 @@ public Request.Processor handle(Request request) throws Exception
 
     // for testing only
-    HttpContent.ContentFactory getContentFactory()
+    HttpContent.Factory getContentFactory()
```
Rename to `getHttpContentFactory()`.

lachlan-roberts replied:
done

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceHandler.java:
```diff
@@ -97,12 +105,33 @@ public void doStart() throws Exception
+        contentFactory = new FileMappedHttpContentFactory(contentFactory);
+        contentFactory = new ValidatingCachingContentFactory(contentFactory, Duration.ofSeconds(1).toMillis(), _byteBufferPool);
+        return contentFactory;
+    }
```
Please also add `get/setHttpContentFactory()` so that it's easier to configure before `start()` (does not require subclassing).

lachlan-roberts replied:
I have now created these methods, and `setupHttpContentFactory()` is only called if there was no `HttpContentFactory` set manually.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ValidatingCachingContentFactory.java:
```diff
@@ -0,0 +1,193 @@
+     * @param maxCacheIdleTime amount of time in ms an entry can be unused before evicted by the sweeper (if 0 never evict unused entries).
+     */
+    public ValidatingCachingContentFactory(@Name("authority") HttpContent.Factory authority,
+                                           @Name("validationTime") long validationTime,
```
Rename to `validationPeriod`.

lachlan-roberts replied:
done

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java:
```diff
@@ -0,0 +1,124 @@
+    {
+        HttpContent content = _factory.getContent(path);
+        if (content != null && content.getContentLengthValue() > _minFileSize)
+            return new FileMappedContent(content);
```
So we don't map the buffer here because we are not sure if something will call `getByteBuffer`?
What exactly is the setup that could cause that to happen?

sbordet replied:
We don't map it here because it's a heavy operation that blocks and takes time, which we don't want to spend in the supposedly fast and light `getHttpContent()`.

gregw replied:
This feels so wrong.  The fast path here is for a file that has already been mapped long ago, so we should just be able to use it directly without a lock and probably without even a volatile lookup.

Firstly are we really sure that mapping a file is a heavy operation?  I would have thought it would be cheaper than reading the file as it just makes an entry into the memory map of the process.   But let's assume that it is expensive....

Then the cache/factory should be able to deal with this in one of several ways:
 1. just block and map it.  Let's see what the cost of this slow path is.
 2. return the original HC until such time as the mapped buffer is loaded, at which time the entry in the cache is replaced with the fast mapped HC
 3. return a HC that waits (preferably threadlessly... but that is another issue) until the cache has been updated with the fast mapped HC.

In short, we need to stop putting policy into `getByteBuffer` that is based on a-priori  knowledge of how it is probably being used.  The `getByteBuffer` method on a `FileMappedHttpContent` should just return the mapped file buffer and leave the policy optimization to be done elsewhere.

lachlan-roberts replied:
Note that when an `HttpContent` is cached it is wrapped by `CachedHttpContent` which will extract out the buffer using `getByteBuffer()`. It is then stored as a final field in `CachedHttpContent` and then the usage of this cached version never hits the volatile just the final field.

If you do not use the `CachingHttpContentFactory`, then using `FileMappedHttpContent.getBuffer()` directly will be hitting this volatile and potentially locking to generate the mapped ByteBuffer.

gregw replied:
Good point, except now we have just transferred the delay/work from the `getByteBuffer` call on the FMHC to the `getHttpContent` on the CCF.    I.e any work done in getting the buffer in a HC that might be cached will end up being work done in the `getHttpContent` on the cache, so we have the same global lock and/or stampede problems there.

Note that I'm a bit dubious if we should ever map a file that is not being cached, as it is a big investment to make for a once off file.

I'm starting to think that file mapping should actually just be an aspect of the cache, that when it decides to load a resource it can do so either as a mapped file or as a byte buffer.  The cache entry can initially be one that either blocks; or calls the non cached HC;  or eventually queues requests until the real HC is ready and replaces it in the cache.

sbordet replied:
File mapping should not be a cache issue. 
If I want to map a file because an external process also maps it and changes it, I would like to return a mapped buffer even if there is no cache in the front.

lachlan-roberts replied:
> we have just transferred the delay/work from the getByteBuffer call on the FMHC to the getHttpContent on the CCF. I.e any work done in getting the buffer in a HC that might be cached will end up being work done in the getHttpContent on the cache, so we have the same global lock and/or stampede problems there.

Because we do the file mapping in `getByteBuffer()` we don't have a stampede of threads all mapping the file. 

Each thread gets a `FileMappedHttpContent` from the factory, they all try to add them to the cache, but the creation of the `CachingHttpContent` (which calls `getByteBuffer()`) is done atomically inside the `_cache.computeIfAbsent` so only one thread will ever do the buffer mapping of the file, the rest of the `FileMappedHttpContent` instances will be discarded because they did not win the race to add it into the cache and they will not also map the file because `getByteBuffer()` was never called.

gregw replied:
Maybe, but that would be yet another case where the implementation of `getByteBuffer` has to somehow telepathically know the intentions of the caller  - does it really want a mapped buffer or not?  is it calling it to cache it or not?  It's just another indication that we don't have the API exactly right yet... together with the fact that we've just had several cycles debating if it is best to do the work in the getContent or the getBuffer, only to realise that if it is cached then all the work is done in the getContent anyway!    We just can't tell why `getByteBuffer` is being called, so we can't optimise it or apply any reasonable policy that holds generally.  

There is a lot of goodness in this PR and I think the things I don't like are due to the API not being correct yet.   I think we need to get the design right (in another PR) and then focus back in on micro optimising the implementations.  So let's get this merged and then consider the API.

gregw replied:
> Each thread gets a `FileMappedHttpContent` from the factory, they all try to add them to the cache, but the creation of the `CachingHttpContent` (which calls `getByteBuffer()`) is done atomically inside the `_cache.computeIfAbsent` so only one thread will ever do the buffer mapping of the file, the rest of the `FileMappedHttpContent` instances will be discarded because they did not win the race to add it into the cache and they will not also map the file because `getByteBuffer()` was never called.

And this is yet another example of a fragile design.   The cache is only not stampeding because we "known" that all the expensive stuff is done in `getByteBuffer`.

I'd much rather a cache design where the element initially put into the cache could make an explicit choice when other requests arrived for the same resource to block them waiting; queue them asynchronously; or dispatch them to the uncached content.     We would have explicit behavior encoded rather than emergent behavior from disciplined implementations of cheap constructors and expensive getBuffers.

sbordet replied:
> Maybe, but that would be yet another case where the implementation of getByteBuffer has to somehow telepathically know the intentions of the caller

That is not true. The implementation does not care at all. It just returns a mapped buffer.
Whether it is the same one mapped once, or many mapped multiple times, the caller does not know and does not care.
You can argue that it is not efficient to map multiple times, but it has nothing to do with the API design.

> We just can't tell why getByteBuffer is being called, so we can't optimise it or apply any reasonable policy that holds generally.

I still don't understand what you are talking about here.
Not a single implementation cares about why or when `getByteBuffer()` is being called.
It is called, it returns a `ByteBuffer` if it can, done.

I still think we are looking for problems that don't exist?

> I'd much rather a cache design where the element initially put into the cache could make an explicit choice when other requests arrived for the same resource to block them waiting; queue them asynchronously; or dispatch them to the uncached content.

We have said repeatedly that this is a better job for a 3rd party cache implementation, not us.

gregw replied:
> > Maybe, but that would be yet another case where the implementation of getByteBuffer has to somehow telepathically know the intentions of the caller
> 
> That is not true. The implementation does not care at all. It just returns a mapped buffer. 

Exactly - a "mapped buffer" because it some how has to know that "an external process also maps it and changes it".

We should just have a HC that serves content as best it can.  If the caller wants a more efficient mechanism, then it is their responsibility to buffer that content or perhaps create a MappedFIleHC.

 
> We have said repeatedly that this is a better job for a 3rd party cache implementation, not us.

Except that we don't have an API that is powerful enough to use a 3rd party cache.  We couldn't use anything that needs to look at cache-control headers for example, as they are not available in getByteBuffer?


I don't understand why you are fighting against this proposal so hard, it can do everything that the getByteBuffer style can and I'm not proposing that initially we change much from what we are currently doing.  But by having a fully powerful API, it allows us to do more powerful things in the future.   It also means that we leverage our existing produce API rather than invent yet another API?

gregw replied:
Closing this discussion to move it elsewhere and not distract from this PR.

lachlan-roberts committed:
changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
add getters and setters for HttpContentFactory onto ResourceHandler

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
use double checked locking for FileMappedHttpContent

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix javadoc

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpContent.java:
```diff
@@ -78,4 +84,136 @@
+    /**
+     * HttpContent Wrapper.
+     */
+    class HttpContentWrapper implements HttpContent
```
Just call it `Wrapper` like all the others.

lachlan-roberts replied:
done

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -61,31 +67,60 @@ public ResourceHandler()
-        setupContentFactory();
-
+        _byteBufferPool = getByteBufferPool(context);
+        if (_resourceService.getHttpContentFactory() == null)
+            _resourceService.setHttpContentFactory(setupHttpContentFactory());
```
```suggestion
        if (getHttpContentFactory() == null)
            setHttpContentFactory(setupHttpContentFactory());
```

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ResourceHandlerTest.java:
```diff
@@ -655,7 +660,19 @@ public void before() throws Exception
+                contentFactory = new ValidatingCachingHttpContentFactory(contentFactory, 0, _local.getByteBufferPool());
+                return contentFactory;
+            }
+        };
```
Use `ResourceHandler.setHttpContentFactory(...)` instead of subclassing.

gregw replied:
Is there a reason to prefer one technique over the other?   

They are slightly different in that subclassing will create the HCF during start, when all configuration is set/known, and that it will then remove that instance during stop, so a new instance is created on next start cycle, using latest configuration.    The setter will set a HCF that will be durable over stop/start cycles, so it will not see config changes in the RH.

So I kind of prefer the subclass style when the configuration of the HCF is entirely dependent on the RH.  The setter style is good for when the HCF has other external configuration.

We should have tests that use both style to make sure they work.   In these specific tests, it doesn't really matter one way or the other.

lachlan-roberts replied:
After a hangout with Simone we decided to not add the setter method for HttpContentFactory at all as it makes the configuration more difficult.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ResourceHandlerTest.java:
```diff
@@ -1941,6 +1957,9 @@ public void testEtagIfNoneMatchModifiedFile() throws Exception
         }
         while (Files.getLastModifiedTime(testFile).equals(before));
 
+        // The cached Resource will only go to fileSystem for expiryTime once per second.
```
`expiryTime` is now called differently?
Also, I would not assume default values. Explicitly configure the period in `ValidatingCachingHttpContentFactory` with a large-ish value (e.g. `period = 2000`), and here wait for `1.5 * period`, so we don't make this test flaky for 100 ms.

lachlan-roberts replied:
Its already configured at 0, just forgot to remove these sleeps.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/TryPathsHandlerTest.java:
```diff
@@ -94,7 +96,16 @@ public void dispose()
+                // We don't want to cache not found entries for this test.
+                return new ResourceHttpContentFactory(ResourceFactory.of(getBaseResource()), getMimeTypes());
+            }
+        };
```
Use the setter now, don't subclass.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/TryPathsHandlerTest.java:
```diff
@@ -160,7 +171,15 @@ public void process(Request request, Response response, Callback callback)
+                // We don't want to cache not found entries for this test.
+                return new ResourceHttpContentFactory(ResourceFactory.of(getBaseResource()), getMimeTypes());
+            }
+        };
```
Use the setter don't subclass.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceHandler.java:
```diff
@@ -97,12 +105,44 @@ public void doStart() throws Exception
 
-        _resourceService.setContentFactory(new ResourceContentFactory(this, _mimeTypes, _resourceService.getPrecompressedFormats()));
+        _byteBufferPool = getByteBufferPool(_context);
+        if (_resourceService.getHttpContentFactory() == null)
+            _resourceService.setHttpContentFactory(setupHttpContentFactory());
```
```suggestion
        if (getHttpContentFactory() == null)
            setHttpContentFactory(setupHttpContentFactory());
```

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java:
```diff
@@ -595,6 +596,39 @@ public static void readFrom(Path path, ByteBuffer buffer) throws IOException
+        int pos = BufferUtil.flipToFill(byteBuffer);
+        try
+        {
+            int totalRead = 0;
```
Should be a `long`.

gregw replied:
I don't think so as it is size limited by the ByteBuffer, which can only be int sized.

lachlan-roberts replied:
done

lachlan-roberts committed:
fixes to the CachingHttpContentFactory caching algorithm

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
add return to BufferUtil.readFrom javadoc

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-HttpContent-Caching-Refactor

lachlan-roberts committed:
remove setter for HttpContentFactory on ResourceHandler
revert to using computeIfAbsent on CachingHttpContentFactory
make direct buffers configurable on CachingHttpContentFactory

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix to ResourceHandler for ResourceHandlerTest.testDirectoryOfCollections

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented:
@lachlan-roberts @sbordet @lorban can we all work together to land this one.  It has been outstanding too long.

gregw reviewed (approved):
I'm OK for t his to be merged as an interim step.
